lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -577,6 +583,63 @@ CompRegCanonicalizer::matchAndRewrite(seq::CompRegOp op,
   return success();
 }
 
+LogicalResult
+MergeVectorizeOps::matchAndRewrite(VectorizeOp op,
+                                   PatternRewriter &rewriter) const {
+  auto operandsVector = op.getOperands();
+  auto groupSize = op.getResults().size();
+
+  SmallVector<SmallVector<Value>> operandsGroups;
+  unsigned counter = 0;
+  SmallVector<Value> temp;
+  for (auto operand : operandsVector) {
+    if (counter == groupSize){
+      operandsGroups.push_back(temp);
+      counter = 0;
+      temp.clear();
+    }
+
+    temp.push_back(operand);
+    counter++;
+  }
+
+  for (auto group : operandsGroups) {
+    auto defOp = group[0].getDefiningOp();
+    bool allFromTheSameGroup = true;
+    SmallPtrSet<Value, 32> uniqueOperands;
+    for (auto operand : group) {
+      if (operand.getDefiningOp() != defOp) {
+        allFromTheSameGroup = false;
+        break;
+      }
+
+      if (!operand.hasOneUse())
+        return failure();
+
+      uniqueOperands.insert(operand);
+    }
+
+    if (allFromTheSameGroup) {
+      if (uniqueOperands.size() != groupSize)
+        return failure();
```

Comment:
Nit, but will `uniqueOperands` ever not just be a copy of groups? As far as I can tell, the only case in which they would be different is where you break from the for loop over `group` early, in which case you'll return failure anyway as the size won't equal groupSize, so you could avoid a little bit of overhead with
```suggestion
    for (auto operand : group) {
      if (operand.getDefiningOp() != defOp)
        return failure();

      if (!operand.hasOneUse())
        return failure();
    }
```
or maybe this is just for type conversion?

Reply:
Thank you very much for the review.
> Will `uniqueOperands` ever not just be a copy of groups?

It will be a subset of `groups` of course but I need to make sure that all the operands of the definedOp appear in this operation and only have one use because this is one of the conditions to merge ops (This is an initial thoughts)

Reply:
You can probably very quickly check whether one of the vector inputs is defined by another vector by comparing the `OperandRange`s:
```
// This gives you an OperandRangeRange, i.e. a 2D array of operands.
// The outer dimension corresponds to arguments to the body block.
// The inner dimension corresponds to individual vector elements.
auto inputVecs = vecOp.getInputs();

// This iterates over the vectors flowing into the vectorize op:
for (auto inputVec : inputVecs) {
  // inputVec is an OperandRange, where each element corresponds to a vector element.
  // You can pick the first element, get its defining op, check if it's a vectorize op, and check if
  // all elements in `inputVec` are exactly the elements from that other vectorize op.
  auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
  if (!otherVecOp) continue;
  if (inputVec != otherVecOp.getResults()) continue;

  // Now you know that inputVec matches the results of otherVecOp exactly.
  // Now you can check if the current vecOp is the sole user of otherVecOp.
  if (!llvm::all_of(otherVecOp.getResults(), [](auto result) { return result.hasOneUse(); }))
    continue;

  // At this point you know that inputVec is exactly the result from otherVecOp, and
  // that this is the only use of otherVecOp. Now you should be able to merge the two.
}
```